tools/SourceKit/lib/SwiftLang/SwiftASTManager.h

```diff
@@ -68,24 +132,97 @@ class ASTUnit : public SourceKit::ThreadSafeRefCountedBase<ASTUnit> {
 
 typedef IntrusiveRefCntPtr<ASTUnit> ASTUnitRef;
 
-class SwiftASTConsumer {
+class SwiftASTConsumer : public std::enable_shared_from_this<SwiftASTConsumer> {
+  /// Mutex guarding all accesses to CancellationRequestCallback
+  llvm::sys::Mutex CancellationRequestCallbackMtx;
+
+  /// A callback that informs the \c ASTBuildOperation, which is producing the
+  /// AST for this consumer, that the consumer is no longer of interest. Calling
+  /// this callback will eventually call \c cancelled on this consumer.
+  /// If the consumer isn't associated with any \c ASTBuildOperation at the
+  /// moment (e.g. if it hasn't been scheduled on one yet or if the build
+  /// operation has already informed the ASTConsumer), the callback is \c None.
+  Optional<std::function<void(std::shared_ptr<SwiftASTConsumer>)>>
+      CancellationRequestCallback;
+
 public:
   virtual ~SwiftASTConsumer() { }
+
+  // MARK: Cancellation
+
+  /// The result of this consumer is no longer of interest to the SourceKit
+  /// client.
+  /// This will cause \c cancelled to be called on this \c SwiftASTConsumer and
+  /// cause the \c ASTBuildOperation to be cancelled if no other consumer is
+  /// depending on it.
+  void requestCancellation() {
+    CancellationRequestCallbackMtx.lock();
+    if (CancellationRequestCallback.hasValue()) {
+      // Copy the function to a new variable so we can call it thread-safe
+      // without CancellationRequestCallbackMtx being locked. This allows the
+      // callback to remove itself from this SwiftASTConsumer.
```

Comment:
Why not remove the callback immediately? Then you don't need to have the callback do it explicitly, which is going to immediately re-take this same lock.  It also ensures that cancellation is idempotent from the perspective of the consumer.

Reply:
Good question. Sounds like a good idea.